### ISSUE TITLE
Issue Report - Segmentation Fault and Browser Compatibility Problem with New TLS Certificates

### ISSUE THREAD

**Issue by sahayap** [michaelrsweet/pappl#289]
Dear Michael,

I hope this message finds you well. We are actively developing AirPrint support for our label printer using pappl-1.3.2, and our progress has generally been positive. However, we have encountered a critical issue regarding the creation of New TLS Certificates  via the Web Interface.

**Problem 1 - Segmentation Fault:**

When attempting to generate a new certificate using the **"Create New Certificate"** button, we consistently **encounter a segmentation fault**. Our investigation into the matter led us to identify the root cause, which lies within the **tls_make_certificate function** in the **system-webif.c** file.

Specifically, the following line of code  is causing the segmentation fault:

`if (!X509_EXTENSION_create_by_NID(&san_ext, NID_subject_alt_name, 0, san_asn1))`

Upon analysis, we found that the variable **san_ext was not properly initialized to NULL**. This oversight resulted in unexpected behavior. We have resolved the segmentation fault by explicitly **initializing san_ext to NULL** and then freeing and **reassigning it to NULL after use,** as shown in the code snippet below:

```
X509_EXTENSION *san_ext = NULL; // Revised initialization - Change 1
// ...
for (i = 0; i < num_alt_names; i++) {
    // The subjectAltName value for DNS names starts with a DNS: prefix...
    snprintf(dns_name, sizeof(dns_name), "DNS:%s", alt_names[i]); // Note Removed the Space between DNS: %s - Change 2

    if ((san_asn1 = ASN1_OCTET_STRING_new()) == NULL)
        break;

    ASN1_OCTET_STRING_set(san_asn1, (unsigned char *)dns_name, (int)strlen(dns_name));
    if (!X509_EXTENSION_create_by_NID(&san_ext, NID_subject_alt_name, 0, san_asn1)) {
        ASN1_OCTET_STRING_free(san_asn1);
        break;
    }
    X509_add_ext(cert, san_ext, -1);
    X509_EXTENSION_free(san_ext);
    san_ext = NULL; // Freed and reassigned to NULL - Change 3
    ASN1_OCTET_STRING_free(san_asn1);
}


```
This 3 changes resolves the segmentation fault issue.

**Problem 2 - Browser Compatibility:**

However, we've noticed that certificates generated using **this approach are only compatible with Safari and not with Firefox, Edge, and Chrome.** 

Our current setup utilizes **openssl-1.1.1g alongside pappl-1.3.2**.

**Request for Assistance:**

We are seeking your expertise and guidance on how to address the compatibility issue with different browsers. Your assistance in resolving this concern would be greatly appreciated.

Thank you for your time and support.

With Regards,
Sahaya Darcius P

**Comment by sahayap** [comment]
Dear Michael,

I've reviewed the OpenSSL test files and successfully discovered a method to generate SubjectAltNames. I've made the necessary modifications to the system_webif.c file, and it now functions seamlessly across all web browsers.

Attached, you'll find the patch file for system_webif.c version 1.3.2. Kindly review it, and if it meets your approval, kindly apply the patch.

[openssl_support_patch.zip](https://github.com/michaelrsweet/pappl/files/12509617/openssl_support_patch.zip)

Thank you for your attention and assistance.

Best regards,
Sahaya Darcius P
